ci(aio) deployment mode features

petebacondarwin opened the pull request:
Closes #18287 
Closes #18189

- [x] deployment to firebase
- [x] mode activation (via environment const or URL query param)
- [x] `aio-shell` element css class (`mode-{archive/latest/next}`)
- [x] info banner for archive mode
- [x] add robots.txt for archive mode
- [x] redirect marketing pages
- [x] adjust version dropdown menu
- [x] CSS style themes

petebacondarwin committed:
feat(aio): add "archive" and "next" color themes

gkalpak commented on aio/src/app/app.component.html:
```diff
@@ -21,8 +21,8 @@
     <aio-nav-menu *ngIf="!isSideBySide" [nodes]="topMenuNarrowNodes" [currentNode]="currentNodes?.TopBarNarrow" [isWide]="false"></aio-nav-menu>
     <aio-nav-menu [nodes]="sideNavNodes" [currentNode]="currentNodes?.SideNav" [isWide]="isSideBySide"></aio-nav-menu>
 
-    <div class="doc-version" title="Angular docs version {{currentDocVersion?.title}}">
```
OOC, why?

petebacondarwin replied:
This just seems wrong. The `div` is a container that holds the drop-down to select versions. Why should it have a `title` that describes just one of the versions?
If anything I would expect it to have a title that said something like `"Select a version of Angular"`.

Moreover, because there were no other `title` attributes, further down, it meant that this tooltip pops up when you hover over any of the individual versions in the drop-down.

![image](https://user-images.githubusercontent.com/15655/28634104-0d9849e6-722e-11e7-9493-35e3b4b7f425.png)

![image](https://user-images.githubusercontent.com/15655/28634121-1bdbff16-722e-11e7-8d66-db34e910f9ce.png)

IgorMinar replied:
was this done for (wrong) accessibility reasons? AFAIK the title attribute is not useful in this context.

petebacondarwin replied:
I guess that was the case.

gkalpak commented on aio/src/app/app.component.ts:
```diff
@@ -129,6 +129,10 @@ export class AppComponent implements OnInit {
     });
 
     this.locationService.currentPath.subscribe(path => {
+      // Redirect to docs if we are in archive mode and not hitting a docs page (i.e. a marketing page)
```
_we are in archive mode_ --> _we are not in stable mode_/_we are in archive or next mode_

gkalpak commented on aio/src/app/app.component.spec.ts:
```diff
@@ -765,6 +765,33 @@ describe('AppComponent', () => {
+        initializeTest();
+        expect(TestBed.get(LocationService).replace).not.toHaveBeenCalled();
+      });
+    })
```
Missing `;` 🙀

petebacondarwin replied:
OMG!

gkalpak commented on aio/src/app/app.component.spec.ts:
```diff
@@ -765,6 +765,33 @@ describe('AppComponent', () => {
+        createTestingModule('docs', 'archive');
+        initializeTest();
+        expect(TestBed.get(LocationService).replace).not.toHaveBeenCalled();
+      });
```
No tests for other modes 😞

petebacondarwin replied:
gotcha!

petebacondarwin replied:
added

petebacondarwin commented on aio/.angular-cli.json:
```diff
@@ -31,8 +31,9 @@
       "environments": {
         "dev": "environments/environment.ts",
-        "stage": "environments/environment.stage.ts",
-        "prod": "environments/environment.prod.ts"
```
OK

petebacondarwin commented on aio/src/environments/environment.ts:
```diff
@@ -14,5 +14,6 @@ import 'core-js/es7/reflect';
   gaId: 'UA-8594346-26', // Staging site
-  production: false
+  production: false,
+  mode: 'stable'
```
I think we agreed that dev env would just use this `gaId` since we can filter out `localhost` requests??  @wardbell or @IgorMinar  might know more.

wardbell replied:
What he said ^^^

petebacondarwin commented on aio/scripts/deploy-to-firebase.sh:
```diff
@@ -3,34 +3,42 @@
+    ;;
   *)
-    echo "Unknown deployment environment ('$deployEnv'). Expected 'staging' or 'production'."
+    echo "Unknown deployment environment ('$deployEnv'). Expected 'next', 'stable' or 'archive'."
```
OK

petebacondarwin commented on scripts/ci/deploy.sh:
```diff
@@ -49,21 +49,17 @@ case ${CI_MODE} in
         exit 0
```
:-( you're right. How about we try to deploy for any build that is at the head of a branch?
The actual firebase deploy will fail because there will be no matching FB project.

For example, building `my-temp-branch` on `angular/angular` would result in `deploy-to-firebase.sh` attempting to deploy to the FB project called `angular-io-my-temp-branch` at the url `https://vmy-temp-branch.angular.io/`

Or is this too hacky?

Otherwise we have to somehow work out if the branch is a valid archive branch or not...

gkalpak replied:
But archive branches will not deploy to a project matchig their name exactly, would they? Because we only care about the major version.
We need to somehow detect if a branch is an archive branch and even further detect if it is the latest branch for that major version. E.g. we don't want to deploy to `v4.angular.io` from `4.2.x` once `4.3.x` has been released.

petebacondarwin replied:
Agh! Yes we don't want to deploy from 4.2.x at all once 4.3.x is released...
What to do ...

petebacondarwin replied:
So instead when we have an archive release, we search github for the most recent minor release for that archive's major version and only deploy if that matches the current branch, right?

petebacondarwin replied:
I updated the code for this... PTAL

petebacondarwin committed:
feat(aio): redirect marketing pages to `docs` if deploy mode is `archive`

See #18287

petebacondarwin committed:
feat(aio): update UI based on deployment mode

* Add a banner if the mode is "archive"
* Add a `mode-...` class to the `aio-shell` element to enable
mode based theming.

See #18287

petebacondarwin committed:
feat(aio): add deploy mode to version picker

See #18287

petebacondarwin committed:
ci(aio): include extra files in AIO deployment based on mode

Any files that are inside the `extra-files/{mode}` folder
will be copied over to the `dist` folder before deployment
to Firebase.

See #18287

IgorMinar reviewed (commented):
There is no preview link - I guess because of the recent CI issues.

Overall this looks very good. Posting a few minor comments. Let's wait for the preview before merging this in.

mary-poppins commented:
You can preview 7f4d13b at https://pr18297-7f4d13b.ngbuilds.io/.

petebacondarwin commented on aio/src/app/layout/mode-banner/mode-banner.component.ts:
```diff
@@ -0,0 +1,16 @@
+  selector: 'aio-mode-banner',
+  template: `
+  <div *ngIf="mode == 'archive'" class="mode-banner">
+    This is the <strong>archived documentation for Angular v{{version?.major}}.</strong>
```
I'll check

petebacondarwin replied:
No, the version info is set via a subscription in the AppComponent: https://github.com/angular/angular/blob/205abe814017a48ec602db959dea37966292dd32/aio/src/app/app.component.ts#L176

petebacondarwin commented on aio/src/extra-files/archive/robots.txt:
```diff
@@ -0,0 +1,2 @@
+User-agent: *
+Disallow: /
```
done

petebacondarwin commented on aio/src/app/shared/deployment.service.ts:
```diff
@@ -1,7 +1,17 @@
+   * The deployment mode set from the environment provided at build time;
+   * or overridden by the `deploymode` query parameter: e.g. `...?deploymode=archive`
+   */
+  mode: string = this.location.search()['deploymode'] || environment.mode;
```
I was worried that 'mode' would be too generic and might conflict with some future feature...
Since this is a debug thing, how about '_mode'?

petebacondarwin replied:
Actually let's just go with 'mode' - first come first served

petebacondarwin committed:
feat(aio): enable deployment mode to be set via URL query

The deployment mode set from the environment provided at build time;
or overridden by the `mode` query parameter: e.g. `...?mode=archive`

See #18287

petebacondarwin commented:
Updated @IgorMinar - PTAL

mary-poppins commented:
You can preview 771a433 at https://pr18297-771a433.ngbuilds.io/.

petebacondarwin commented:
Some examples:

Archive mode: https://pr18297-771a433.ngbuilds.io/tutorial?mode=archive

![screen shot 2017-07-27 at 07 14 12](https://user-images.githubusercontent.com/15655/28656686-40aaf848-729b-11e7-89d7-d19a3f7ebff9.png)

Next mode: https://pr18297-771a433.ngbuilds.io/tutorial?mode=archive

![screen shot 2017-07-27 at 07 15 00](https://user-images.githubusercontent.com/15655/28656717-6ab212f2-729b-11e7-9ef5-ba2536e78d7d.png)

Redirection - clicking on this link will redirect you to the docs page: https://pr18297-771a433.ngbuilds.io/resources?mode=next. Note that the query string doesn't survive the redirection (which is not a problem in practice) but does mean that the mode disappears after the redirection in this case.

petebacondarwin commented on aio/package.json:
```diff
@@ -9,7 +9,7 @@
     "start": "yarn check-env && ng serve",
     "prebuild": "yarn check-env && yarn setup",
-    "build": "ng build -prod -sm -vc=false",
+    "build": "ng build ng build --target=production --environment=stable -sm -vc=false",
```
This change fixes @gkalpak comment https://github.com/angular/angular/pull/18297/files#r129411614

mary-poppins commented:
You can preview 377cf4b at https://pr18297-377cf4b.ngbuilds.io/.

gkalpak commented on aio/scripts/deploy-to-firebase.sh:
```diff
@@ -3,41 +3,77 @@
+  readonly majorVersion=${TRAVIS_BRANCH%%.*}
+  readonly mostRecentMinorVersion=$(
+    # List the branches that start with the major version
+    git ls-remote https://github.com/angular/angular refs/heads/${majorVersion}.*.x |
```
Nit: Why not use `origin`? (We already use it to determine the lastest commit.)

petebacondarwin replied:
Because I was worried that `origin` might not exist on a CI server. But I guess, if we are using it elsewhere then it is less fragile for using with a different repo.

gkalpak commented on aio/scripts/deploy-to-firebase.sh:
```diff
@@ -3,41 +3,77 @@
+    # Get the highest version
+    tail -n1)
+  if [ $TRAVIS_BRANCH == $mostRecentMinorVersion ]; then
+    readonly deployEnv=archive
```
This will still deploy `4.2.x` to `v4.angular.io` when `4.3.x` is the stable. Is it intentional?

petebacondarwin replied:
While 4.3.x is stable, there will be no `angular-io-4` firebase project, so it would just fail.
Perhaps we want not to fail but just quietly not deploy...

petebacondarwin replied:
OK, so I'll change it so that we quietly fail if there is a branch with the same major but a higher minor version number - even if that higher branch is marked as the stable branch.

gkalpak commented on aio/scripts/deploy-to-firebase.sh:
```diff
@@ -3,41 +3,77 @@
-    readonly buildEnv=stage
+  next)
     readonly projectId=aio-staging
     readonly deployedUrl=https://$projectId.firebaseapp.com/
```
Shouldn't this be `https://next.angular.io/`?

petebacondarwin replied:
Hmm, well strictly we are deploying to `aio-staging.firebaseapp.com` it just happens that this is also `next.angular.io`. Happy to change later if that is a problem.

gkalpak replied:
Strictly we are deploying stable to `angular-io.firebaseapp.com`, but that didn't stop you from use `https://angular.io` as the deployed URL 😛 

There shouldn't be any practical difference (except maybe that using `next.angular.io` would also verify that the DNS is configured correctly 🎉).

gkalpak commented on aio/scripts/deploy-to-firebase.sh:
```diff
@@ -3,41 +3,77 @@
+  yarn build -- --env=$deployEnv
+
+  # Include any mode-specific files
+  cp -rf src/extra-files/$buildEnv/. dist/
```
`$buildEnv` --> `$deployEnv`

petebacondarwin replied:
:scream:

This file really should have some kind of unit test :-(

petebacondarwin commented on aio/src/environments/environment.ts:
```diff
@@ -13,6 +13,7 @@ import 'core-js/es7/reflect';
 export const environment = {
-  gaId: 'UA-8594346-26', // Staging site
-  production: false
+  gaId: 'UA-8594346-26', // Staging id
```
OK - let's call it dev

petebacondarwin committed:
ci(aio): compute AIO deployment mode

There are now 3 modes for deployment: next, stable, archive.
We compute which mode (and other deployment properties)
from the `TRAVIS_BRANCH` and the `STABLE_BRANCH`.

If the TRAVIS_BRANCH is master we deploy as "next".
Otherwise if the branch is the highest of its minor versions
we deploy as "stable" if the `TRAVIS_BRANCH` matches the `STABLE_BRANCH` or
else "archive".

For "archive" deployments we compute the firebase project and deployment
url based on the major version of the `TRAVIS_BRANCH`.

As well as choosing where to deploy the build, we also use this
to select the environment file for the AIO Angular app.
This will enable the app to change its rendering and behaviour
based on its mode.

See #18287

petebacondarwin commented on aio/scripts/deploy-to-firebase.sh:
```diff
@@ -3,41 +3,83 @@
+    echo "Skipping deploy of branch \"${TRAVIS_BRANCH}\" to firebase."
+    echo "There is a more recent branch with the same major version: \"${mostRecentMinorVersion}\""
+    exit 0
+  fi
```
So now we will only deploy to `stable` if there is no branch with a the same major and higher minor version. This means that we couldn't have a stable branch of `4.3.x` and then have another branch with `4.4.x`.

gkalpak replied:
Hm...that sound almost right, but...
What happens when 4.3.x is the stable branch and we want to deploy `4.4.0-beta.1` from the 4.4.x branch?
We need a more elaborate check that only archives branches below the current stable branch.

mary-poppins commented:
You can preview 58aa117 at https://pr18297-58aa117.ngbuilds.io/.

petebacondarwin committed:
ci(aio): compute AIO deployment mode

There are now 3 modes for deployment: next, stable, archive.
We compute which mode (and other deployment properties)
from the `TRAVIS_BRANCH` and the `STABLE_BRANCH`.

If the TRAVIS_BRANCH is master we deploy as "next".
If the `TRAVIS_BRANCH` matches the `STABLE_BRANCH` we deploy as "stable".

Otherwise if the branch has a major version lower than the stable version
and its minor version is highest of similar branches we deploy as "archive".

For "archive" deployments we compute the firebase project and deployment
url based on the major version of the `TRAVIS_BRANCH`.

As well as choosing where to deploy the build, we also use this
to select the environment file for the AIO Angular app.
This will enable the app to change its rendering and behaviour
based on its mode.

See #18287

mary-poppins commented:
You can preview e668e8b at https://pr18297-e668e8b.ngbuilds.io/.

gkalpak commented on aio/scripts/deploy-to-firebase.test.sh:
```diff
@@ -0,0 +1,135 @@
+function check {
+  if [[ $1 == $2 ]]; then
+  echo Pass
+  exit 0
```
No indentation 😞

gkalpak commented on aio/scripts/deploy-to-firebase.test.sh:
```diff
@@ -0,0 +1,135 @@
+(
+  echo ===== archive - skip deploy - major version too high
+  actual=$(
+    export TRAVIS_BRANCH=2.1.x
```
I might add a test with `TRAVIS_BRANCH.minor > STABLE_BRANCH.minor` (e.g. `TRAVIS_BRANCH=2.4.x`, `STABLE_BRANCH=2.2.x`) just to make it more obvious.
(It is essentally the same check that is in effect for this test, it just makes the intent in that specific corner-case more obvious.)

gkalpak reviewed (approved):
Love the bash script unit test. It's a shame not to run it on CI, though 😞 

LGTM (as soon as Travis is green).

gkalpak commented on scripts/ci/deploy.sh:
```diff
@@ -39,34 +40,14 @@ case ${CI_MODE} in
       )
       travisFoldEnd "deploy.aio"
-
     fi
```
Oh, this `fi` should go (kudos to Travi for catching that).

mary-poppins commented:
You can preview a16e9dc at https://pr18297-a16e9dc.ngbuilds.io/.

mary-poppins commented:
You can preview 25e25d9 at https://pr18297-25e25d9.ngbuilds.io/.

mary-poppins commented:
You can preview 0856320 at https://pr18297-0856320.ngbuilds.io/.

alxhub commented:
Removing the merge label as this doesn't have full approval yet. I'll ping it to Igor.

vicb commented:
conflicts -> blocked
@petebacondarwin could you please rebase ?

mary-poppins commented:
You can preview 68ad94c at https://pr18297-68ad94c.ngbuilds.io/.

gkalpak commented:
Rebased. (CI failures "inherited" from master.)

vicb commented:
Thanks for rebasing @gkalpak
